Improve common student error messages

microbit-robert opened the pull request:
Add a new simplified message bundle with simplified text of error message after review with the education team.

In simplified mode, types aren't referenced in common messages as students in our scenarios are best focussed on docs and examples to fix their errors.

New messages:
- booleanIsLowercase special-cases "true" and "false" in name errors as forgetting to uppercase is common
- expectedEqualityOperator adds a new case for `if a = b` scenarios

Code-level tweaks to existing messages:
- add module names to importSymbolUnknown and moduleUnknownMember (potentially upstreamable)

microbit-robert committed:
Add and use micro:bit error strings

microbit-robert committed:
Special case true and false

microbit-robert committed:
Special case assignment/equality mistake

microbit-robert committed:
Name module for moduleUnknownMember

microbit-matt-hillsdon committed:
Naming tweak.

microbit-matt-hillsdon commented on packages/pyright-internal/src/localization/package.mb.en.json:
```diff
@@ -0,0 +1,15 @@
+{
+    "Diagnostic": {
+        "booleanIsLowerCase": "\"{name}\" is not defined, did you mean \"{booleanName}\"?",
```
Let's add the common errors that we didn't change the text of, as we want them to all be in a single file for translation purposes.

microbit-robert replied:
This seems fairly arbitrary on my part, but these are the errors I commonly see - [`edd1ad4`](https://github.com/microbit-foundation/pyright/pull/10/commits/edd1ad4c1bf21e70c8e633a352832e51836162da).

microbit-matt-hillsdon replied:
Let's discuss the text with Giles. Please can you add some more context for the "Expected expression" scenarios? Perhaps it is just very general but would be good to confirm.

microbit-matt-hillsdon commented on packages/pyright-internal/src/parser/parser.ts:
```diff
@@ -1207,10 +1207,17 @@ export class Parser {
-
+            if (nextToken.type === TokenType.Operator) {
+                comparisonOperator = this._peekOperatorType();
+                if (comparisonOperator === OperatorType.Assign) {
```
Can we just inline this._peekOperatorType() and remove the comparisonOperator var?

There are some tests in parser.test.ts that could be extended with a test to cover this with an example like:

```
a = 0
if a = 1:
    pass
```

microbit-robert replied:
Inline code - [`45d87ec`](https://github.com/microbit-foundation/pyright/pull/10/commits/45d87ecaeb9f1945275662bd0f7cd4bb42ecbea6).
Add tests - [`28a8cff`](https://github.com/microbit-foundation/pyright/pull/10/commits/28a8cffcaecc4d402204dd6a88044a52787a2525).

microbit-robert committed:
Tidy

microbit-robert committed:
Add more common strings

microbit-robert committed:
Simplify arg/param type errors

microbit-robert committed:
Simplify invalid operators on types error messages

microbit-robert committed:
Add other common error strings to mb file

microbit-robert committed:
Add tests for equality operator error

microbit-robert commented:
Currently failing `localizer.test.ts` for having duplicate strings for different keys in our custom strings JSON file.

We can change the strings slightly, make code changes to return only one error key to avoid the duplication, or disable the code in the test that checks this.

```
if (stringContentMap.has(formatString)) {
    assert.fail(`Localized string for "${key}" is duplicate of ${stringContentMap.get(formatString)}`);
}
```

microbit-robert commented on packages/pyright-internal/src/localization/localize.ts:
```diff
@@ -156,16 +174,26 @@ function loadStringsForLocale(locale: string): StringLookupMap {
         if (override !== undefined) {
             return override;
         }
+        // Fallback to English micro:bit strings.
+        if (microbit) {
+            override = loadStringsFromJsonFile('en', microbit);
+            if (override !== undefined) {
+                return override;
+            }
+        }
```
Explicit fallback is required for tests, though this might not be the desired behaviour where the translation is not available for another language.

microbit-matt-hillsdon replied:
Simpler approach in 7468c84f09be08f8eae31002622c67b54a03d755

microbit-robert commented on packages/pyright-internal/src/localization/package.mb.en.json:
```diff
@@ -0,0 +1,30 @@
+        "expectedCloseParen": "Missing closing bracket \")\"",
+        "expectedColon": "Missing colon \":\"",
+        "expectedEqualityOperator": "Expected equality operator, did you mean \"==\"?",
+        "expectedExpr": "Expected expression",
```
Expected expression occurs where it is missing from:

- if/while statements; e.g. `if :`
- the right hand side of operands excluding assign (`=`) which is special cased

microbit-matt-hillsdon replied:
"excluding assign (=) which is special cased" <- should that be a common error?

microbit-robert committed:
Simplify type not iterable message

microbit-matt-hillsdon committed:
Simplify string override logic.

Add exceptions to the string clash test.

microbit-matt-hillsdon committed:
Merge branch 'improve-errors' of github.com:microbit-foundation/pyright into improve-errors

microbit-matt-hillsdon commented:
@microbit-robert, please can you update the python editor branch with a new build and check all's good?

I think the PR looks good, let's review the strings with Giles then I think it's ready to go.

microbit-robert committed:
Final string tweaks

microbit-matt-hillsdon committed:
Only use simplified messages in browser package.

Avoids undesirtable impact on tests, particularly from messages being
deduplicated as their text/pos ends up the same.

microbit-robert commented on packages/pyright-internal/src/localization/localize.ts:
```diff
@@ -180,7 +178,27 @@ function loadStringsForLocale(locale: string): StringLookupMap {
 
 function loadStringsFromJsonFile(locale: string): StringLookupMap | undefined {
-    return stringMapsByLocale.get(locale);
+    console.log('Init', locale, messageStyle);
```
Remove debug.

microbit-robert commented on packages/browser-pyright/src/browser-server.ts:
```diff
@@ -361,6 +364,10 @@ export class BrowserBackgroundAnalysis extends BackgroundAnalysisBase {
         super(parentPort(), initialData);
+
+        // Hack to enable simplified error messages in the background thread.
+        // Ideally we'd route this from via initialData.
```
Clean up comment slightly by removing `from` or `via`.

microbit-matt-hillsdon replied:
```suggestion
        // Ideally we'd route this via initialData.
```

microbit-matt-hillsdon committed:
Tidy

microbit-matt-hillsdon commented on packages/pyright-internal/src/localization/localize.ts:
```diff
@@ -166,7 +178,27 @@ function loadStringsForLocale(locale: string): StringLookupMap {
 
 function loadStringsFromJsonFile(locale: string): StringLookupMap | undefined {
-    return stringMapsByLocale.get(locale);
+    console.log('Init', locale, messageStyle);
```
```suggestion
```

microbit-matt-hillsdon committed:
Remove debug

microbit-matt-hillsdon commented on packages/pyright-internal/src/analyzer/typeEvaluator.ts:
```diff
@@ -3789,8 +3789,20 @@ export function createTypeEvaluator(importLookup: ImportLookup, evaluatorOptions
                 }
             }
         } else {
+            // Handle the special case of boolean".
```
```suggestion
            // Handle the special case of booleans.
```

microbit-matt-hillsdon committed:
Clarify comment

microbit-matt-hillsdon committed:
Exceptions no longer required.

microbit-matt-hillsdon committed:
Merge branch 'improve-errors' of github.com:microbit-foundation/pyright into improve-errors

microbit-robert committed:
Add another common error string

microbit-matt-hillsdon committed:
We now pass module so update this string.

Mostly just because it makes it easier to track our potentially
upstreamable changes.

microbit-matt-hillsdon committed:
Merge branch 'improve-errors' of github.com:microbit-foundation/pyright into improve-errors

microbit-matt-hillsdon committed:
Reinstate addenda.

microbit-matt-hillsdon commented:
I messed up the merge here so going again on #11.
